contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,549 @@
+  
+  GoUint64 go_tatus = dynamic_lib_->envoyGoEncodeData(
+    s, end_stream ? 1 : 0, reinterpret_cast<uint64_t>(&buffer), buffer.length());
+  state.setFilterState(FilterState::Done);
+  state.doDataList.moveOut(data);
+
+  switch (go_tatus) {
+  case static_cast<GoUint64>(SendDataStatus::SendDataWithTunneling):
+    // send data to upstream with conn not half close
+    upstream_conn_data_->connection().write(data, false);
+    break;
+
+  case static_cast<GoUint64>(SendDataStatus::SendDataWithNotTunneling):
+    // send data to upstream with conn half close
+    upstream_conn_data_->connection().write(data, true);
+    break;
+
+  case static_cast<GoUint64>(SendDataStatus::NotSendData):
+    // not send data to upstream
+    break;
+
+  default:
+    // send data to upstream
+    ENVOY_LOG(error, "encodeData unexpected go_tatus: {}", go_tatus);
+    upstream_conn_data_->connection().write(data, end_stream);
+    break;
+  }
+
+}
+
+// TODO(duxin40): use golang to convert trailers to data buffer.
+void TcpUpstream::encodeTrailers(const Envoy::Http::RequestTrailerMap&) {
+  Buffer::OwnedImpl data;
+  upstream_conn_data_->connection().write(data, true);
+}
+
+void TcpUpstream::readDisable(bool disable) {
+  if (upstream_conn_data_->connection().state() != Network::Connection::State::Open) {
+    return;
+  }
+  upstream_conn_data_->connection().readDisable(disable);
+}
+
+void TcpUpstream::resetStream() {
+  upstream_request_ = nullptr;
+  upstream_conn_data_->connection().close(Network::ConnectionCloseType::NoFlush);
+}
+
+void TcpUpstream::onUpstreamData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream onUpstreamData, data length: {}, end_stream: {}", data.length(), end_stream);
+
+  filter_->initResponse();
+
+  DecodingProcessorState& state = filter_->decoding_state_;
+  state.resp_headers = static_cast<Envoy::Http::RequestOrResponseHeaderMap*>(filter_->resp_headers_.get());
+  Buffer::Instance& buffer = state.doDataList.push(data);
+  auto s = dynamic_cast<processState*>(&state);
+  state.setFilterState(FilterState::ProcessingData);
+
+  GoUint64 go_tatus = dynamic_lib_->envoyGoOnUpstreamData(
+    s, end_stream ? 1 : 0, (*filter_->resp_headers_).size(), (*filter_->resp_headers_).byteSize(), reinterpret_cast<uint64_t>(&buffer), buffer.length());
+
+  state.setFilterState(FilterState::Done);
+  state.doDataList.moveOut(data);
+
+  switch (go_tatus) {
+  case static_cast<GoUint64>(ReceiveDataStatus::ReceiveDataFinish):
+    // send data to downstream
+    end_stream = true;
+    // if go side not set status, c++ side set default status
+    if (!filter_->resp_headers_->Status()) {
+      filter_->resp_headers_->setStatus(static_cast<uint64_t>(HttpStatusCode::Success));
+    }
+    upstream_request_->decodeHeaders(std::move(filter_->resp_headers_), false);
+    upstream_request_->decodeData(data, end_stream);
+    break;
+
+  case static_cast<GoUint64>(ReceiveDataStatus::ReceiveDataContinue):
```

Comment:
seem only full-buffer supported? How do we send data downstream, streaming style?

Reply:
Maybe we need to introduce another status?
Use continue for streaming and `waitdata` for the current?

Reply:
done